airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/state/StateManagerFactory.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.source.relationaldb.state;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.annotations.VisibleForTesting;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.integrations.source.relationaldb.models.DbState;
+import io.airbyte.protocol.models.AirbyteStateMessage;
+import io.airbyte.protocol.models.AirbyteStateMessage.AirbyteStateType;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Factory class that creates {@link StateManager} instances based on the provided state.
+ */
+public class StateManagerFactory {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StateManagerFactory.class);
+
+  /**
+   * Private constructor to prevent direct instantiation.
+   */
+  private StateManagerFactory() {}
+
+  /**
+   * Creates a {@link StateManager} based on the provided state object and catalog.
+   *
+   * @param state The deserialized state.
+   * @param catalog The {@link ConfiguredAirbyteCatalog} for the connector that will utilize the state
+   *        manager.
+   * @param config The connector configuration.
+   * @return A newly created {@link StateManager} implementation based on the provided state.
+   */
+  public static StateManager createStateManager(final List<AirbyteStateMessage> state,
+                                                final ConfiguredAirbyteCatalog catalog,
+                                                final JsonNode config) {
+    if (state != null && !state.isEmpty()) {
+      final AirbyteStateMessage airbyteStateMessage = state.get(0);
```

Comment:
We determine the state manager based on the state type attribute of the first element which makes me feel why is this info being duplicated for each stream then? The current AirbyteStateMessage looks like 
```
    private AirbyteStateMessage.AirbyteStateType stateType;
    @JsonProperty("stream")
    private AirbyteStreamState stream;
    @JsonProperty("global")
    private AirbyteGlobalState global;
    /**
     * (Deprecated) the state data
     * 
     */
    @JsonProperty("data")
    @JsonPropertyDescription("(Deprecated) the state data")
    private JsonNode data;
```
Instead shouldnt the `private AirbyteStreamState stream;` be `List<AirbyteStreamState> streams;`  (just like we have it in the `private AirbyteGlobalState global;`)

Reply:
@subodh1810 I will ping in @cgardens and @benmoriceau here.  I think the intent is that there will be a state message for each stream.  In the other two cases (legacy and global) there is only one state message produces.  Because of the per-stream case, this means that a list of states may be passed to the source on execution.  I agree that this code is a bit wonky and could be remedied by having a protocol version AND having a type/interface that represents the state data.  In that case, the type could be pushed there and it could also include a list of state messages.  Something like:

```
interface ConnectorState {

    StateType type();
    
    List<AirbyteStateMessage> states();
    
}
```

However, we are not there yet.

Reply:
+1 to what @jdpgrailsdev mentioned.